File: netlify/functions/_lib/usage.mjs
import { sha256Hex } from './common.mjs'

const memoryCounters = new Map()

function upstashConfig() {
  const url = process.env.REDIS_REST_URL
  const token = process.env.REDIS_REST_TOKEN
  if (!url || !token) return null
  return { url: url.replace(/\/$/, ''), token }
}

async function upstashIncr(key, ttlSeconds) {
  const cfg = upstashConfig()
  if (!cfg) return null

  const headers = { Authorization: `Bearer ${cfg.token}` }
  const incrRes = await fetch(`${cfg.url}/incr/${encodeURIComponent(key)}`, { headers })
  if (!incrRes.ok) throw new Error('upstash incr failed')
  const incrData = await incrRes.json()
  const current = Number(incrData.result || 0)

  if (current === 1) {
    await fetch(`${cfg.url}/expire/${encodeURIComponent(key)}/${ttlSeconds}`, { headers })
  }

  return current
}

export function makeUsageKey(ip, userAgent, isPro) {
  const day = new Date().toISOString().slice(0, 10).replace(/-/g, '')
  const salt = process.env.USAGE_SALT || 'change-me'
  const raw = `${ip}|${userAgent}|${day}|${salt}`
  const digest = sha256Hex(raw)
  return `usage:${isPro ? 'pro' : 'free'}:${digest}`
}

export async function checkAndIncrementUsage(key, isPro) {
  const freeLimit = Number(process.env.FREE_DAILY_LIMIT || 5)
  const proLimit = Number(process.env.PRO_DAILY_LIMIT || 500)
  const limit = isPro ? proLimit : freeLimit
  const ttl = 60 * 60 * 48

  let used
  try {
    const redisUsed = await upstashIncr(key, ttl)
    if (redisUsed !== null) used = redisUsed
  } catch {
    used = undefined
  }

  if (typeof used !== 'number') {
    const current = (memoryCounters.get(key) || 0) + 1
    memoryCounters.set(key, current)
    used = current
  }

  return { allowed: used <= limit, used, limit }
}

File: backend/usage.py
from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional

from redis import Redis
from redis.exceptions import RedisError


@dataclass
class UsageResult:
    allowed: bool
    used: int
    limit: int


class UsageLimiter:
    def __init__(self) -> None:
        self.limit_free = int(os.getenv("FREE_DAILY_LIMIT", "100"))
        self.limit_pro = int(os.getenv("PRO_DAILY_LIMIT", "500"))
        self.salt = os.getenv("USAGE_SALT", "change-me")
        self.redis: Optional[Redis] = None
        self.mem = {}

        redis_url = os.getenv("REDIS_URL")
        if redis_url:
            self.redis = Redis.from_url(redis_url, decode_responses=True)

    def make_usage_key(self, ip: str, user_agent: str, is_pro: bool) -> str:
        day = datetime.now(timezone.utc).strftime("%Y%m%d")
        raw = f"{ip}|{user_agent}|{day}|{self.salt}"
        digest = hashlib.sha256(raw.encode("utf-8")).hexdigest()
        tier = "pro" if is_pro else "free"
        return f"usage:{tier}:{digest}"

    def check_and_increment(self, key: str, is_pro: bool) -> UsageResult:
        limit = self.limit_pro if is_pro else self.limit_free
        ttl = 60 * 60 * 48

        if self.redis:
            try:
                val = self.redis.incr(key)
                if val == 1:
                    self.redis.expire(key, ttl)
                return UsageResult(allowed=val <= limit, used=int(val), limit=limit)
            except RedisError:
                pass

        current = self.mem.get(key, 0) + 1
        self.mem[key] = current
        return UsageResult(allowed=current <= limit, used=current, limit=limit)
